code42_connector.py

```diff
@@ -154,11 +166,90 @@ def _handle_list_high_risk_employees(self, param, action_result):
+        action_result.add_data(alert)
+        action_result.update_summary(
+            {"username": alert["actor"], "user_id": alert["actorId"]}
+        )
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    @action_handler_for("search_alerts")
+    def _handle_search_alerts(self, param, action_result):
+        username = param.get("username")
+        start_date = param.get("start_date")
+        end_date = param.get("end_date")
+        alert_state = param.get("alert_state")
+
+        if username is None and start_date is None and end_date is None and alert_state is None:
+            return action_result.set_status(
+                phantom.APP_ERROR, "Must supply a search term."
+            )
+
+        if not self._validate_date_range(start_date, end_date):
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                "Start Date and End Date are both required to search by date range.",
+            )
+        try:
+            query = self._build_alerts_query(
+                username, start_date, end_date, alert_state
+            )
+        except ValueError as exception:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                f"Start Date and End Date must be in format YYYY-mm-dd: {exception}",
+            )
+
+        response = self._client.alerts.search(query)
+        action_result.add_data(response.data)
+        action_result.update_summary({"total_count": response["totalCount"]})
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    @action_handler_for("set_alert_state")
+    def _handle_set_alert_state(self, param, action_result):
+        alert_id = param["alert_id"]
+        alert_state = param["alert_state"]
+        note = param.get("note")
+        response = self._client.alerts.update_state(alert_state, [alert_id], note=note)
+        action_result.add_data(response.data)
+        status_message = f"State of alert {alert_id} was updated to {alert_state}"
+        action_result.update_summary({"alert_id": alert_id})
+        return action_result.set_status(phantom.APP_SUCCESS, status_message)
+
     def finalize(self):
         # Save the state, this data is saved across actions and app upgrades
         self.save_state(self._state)
         return phantom.APP_SUCCESS
 
+    def _validate_date_range(self, start_date, end_date):
+        return (not start_date and not end_date) or (
+            start_date is not None and end_date is not None
+        )
+
+    def _build_alerts_query(self, username, start_date, end_date, alert_state):
+        filters = []
+        if username is not None:
+            filters.append(Actor.eq(username))
+        if start_date is not None and end_date is not None:
+            # This will raise a ValueError if dates not in Y-m-d H:M:S format
+            filters.append(
+                DateObserved.in_range(f"{start_date} 00:00:00", f"{end_date} 00:00:00")
```

Comment:
does phantom only let you input calendar dates (no timestamps)? Trying to understand the `00:00:00` here.

Reply:
No Phantom doesn't have a dedicated Date or DateTime input so it's just a string. The requirement just said to allow user to supply a date range so I figured that exposing time-level filtering was more granular than the usecase required. Now that I think about it, security analysts searching for alerts probably *do* require that kind of granularity when building out queries.

If we allow the `start_date` and `end_date` parameters to be entered with the timestamp included, should we require that formatting or also allow them to omit the time and only supply date?

Reply:
it looks like the phantom env has the [python-dateutil](https://pypi.org/project/python-dateutil/) package installed, we could just call `dateutil.parser.parse()` on the strings and if it fails say it's not a recognized date format. Then we don't have to do any manual conversions/parsing.

Reply:
That definitely seems more flexible/safe. The comment that we have in the code right now is also misleading (I'm pretty sure something bad would happen if we included `H:M:S` in the input right now since `00:00:00` would still be concat'd on).

Reply:
That's a great idea Tim. I had found that package when I was first researching this problem but was reluctant to add a new dependency for such a little thing. Much cleaner and more flexible now that we're using the `dateutil` parser.

Reply:
I think the API technically allows any format, right?
We shouldn't make it harder to input dates - what if the date is coming from some other app and it is a longer format? It'd be nice if we could just pass it through to this method without the playbook author needing to worry about it